validator/client/src/main/java/tech/pegasys/teku/validator/client/BeaconProposerPreparer.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2021 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package tech.pegasys.teku.validator.client;
+
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.apache.tuweni.bytes.Bytes32;
+import tech.pegasys.teku.infrastructure.unsigned.UInt64;
+import tech.pegasys.teku.spec.Spec;
+import tech.pegasys.teku.ssz.type.Bytes20;
+import tech.pegasys.teku.validator.api.BeaconPreparableProposer;
+import tech.pegasys.teku.validator.api.ValidatorApiChannel;
+import tech.pegasys.teku.validator.api.ValidatorTimingChannel;
+import tech.pegasys.teku.validator.client.loader.OwnedValidators;
+
+public class BeaconProposerPreparer implements ValidatorTimingChannel {
+  private final ValidatorApiChannel validatorApiChannel;
+  private final ValidatorIndexProvider validatorIndexProvider;
+  private final OwnedValidators validators;
+  private final Spec spec;
+  private final Bytes20 feeRecipient;
+
+  public BeaconProposerPreparer(
+      ValidatorApiChannel validatorApiChannel,
+      ValidatorIndexProvider validatorIndexProvider,
+      Optional<Bytes20> feeRecipient,
+      OwnedValidators validators,
+      Spec spec) {
+    this.validatorApiChannel = validatorApiChannel;
+    this.validatorIndexProvider = validatorIndexProvider;
+    this.validators = validators;
+    this.feeRecipient = feeRecipient.orElse(null);
+    this.spec = spec;
+  }
+
+  @Override
+  public void onSlot(UInt64 slot) {
+    if (spec.atSlot(slot).getExecutionPayloadUtil().isPresent()
+        && slot.mod(spec.getSlotsPerEpoch(slot)).isZero()) {
```

Comment:
just asking my self if it is safe for the case of being the proposer of the first slot in Merge fork. We might start calling this from the prior epoch?

Reply:
I don't see any reason this shouldn't be called prior to the merge - there's no point if the merge fork isn't enabled at all but having the beacon node notified ahead of the merge happening doesn't hurt and means the beacon node is definitely ready when the merge fork kicks in.  So I think I'd actually remove the `getExecutionPayloadUtil().isPresent()` check here.

The beacon node just wouldn't do anything with the information about which validators are running until the merge epoch actually kicks in (and then probably only if terminal total difficulty is reached).

Reply:
oky.
my initial concern was a non-problem: forgot to consider that TTD needs to be reached before this data become necessary.